azure/services/aso/aso.go

```diff
@@ -47,25 +47,31 @@ const (
 	deleteFutureType         = "ASODelete"
 )
 
+// deepCopier is a genruntime.MetaObject with a typed DeepCopy method, usually generated by kubebuilder.
+type deepCopier[T any] interface {
+	genruntime.MetaObject
+	DeepCopy() T
+}
+
 // Service is an implementation of the Reconciler interface. It handles creation
 // and deletion of resources using ASO.
-type Service struct {
+type Service[T deepCopier[T]] struct {
```

Comment:
This looks funky and I admit I can't think of a good way to explain how this works.